pkg/trace/stats/quantile/converter.go

```diff
@@ -0,0 +1,171 @@
+	}
+	for i := range s2.bins {
+		index := int(i)
+		if index >= s1.contiguousBinsOffset && index < s1.contiguousBinsOffset+len(s1.contiguousBins) {
+			continue
+		}
+		if index >= s2.contiguousBinsOffset && index < s2.contiguousBinsOffset+len(s2.contiguousBins) {
+			continue
+		}
+		if _, ok := s1.bins[i]; ok {
+			continue
+		}
+		indexes = append(indexes, index)
+	}
+	sort.Ints(indexes)
+	return indexes
+}
+
+// decodeDDSketch decodes a ddSketch from a protobuf encoded ddSketch.
+// It only supports sketches with positive values.
+func decodeDDSketch(data []byte) (ddSketch, error) {
+	var pb sketchpb.DDSketch
+	if err := proto.Unmarshal(data, &pb); err != nil {
+		return ddSketch{}, err
+	}
+	mapping, err := mapping.FromProto(pb.Mapping)
+	if err != nil {
+		return ddSketch{}, err
+	}
+	if len(pb.NegativeValues.BinCounts) > 0 ||
+		len(pb.NegativeValues.ContiguousBinCounts) > 0 {
+		return ddSketch{}, errors.New("negative values not supported")
+	}
+	return ddSketch{
+		mapping:              mapping,
+		bins:                 pb.PositiveValues.BinCounts,
+		contiguousBins:       pb.PositiveValues.ContiguousBinCounts,
+		contiguousBinsOffset: int(pb.PositiveValues.ContiguousBinIndexOffset),
+		zeros:                int(pb.ZeroCount),
+	}, nil
+}
+
+// DDToGKSketches converts two dd sketches: ok and errors to 2 gk sketches: hits and errors
+// with hits = ok + errors
+func DDToGKSketches(okSketchData []byte, errSketchData []byte) (hits, errors *SliceSummary, err error) {
+	okDDSketch, err := decodeDDSketch(okSketchData)
+	if err != nil {
+		return nil, nil, err
+	}
+	errDDSketch, err := decodeDDSketch(errSketchData)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	hits = &SliceSummary{Entries: make([]Entry, 0, okDDSketch.maxSize())}
+	errors = &SliceSummary{Entries: make([]Entry, 0, errDDSketch.maxSize())}
+	if zeros := okDDSketch.zeros + errDDSketch.zeros; zeros > 0 {
+		hits.Entries = append(hits.Entries, Entry{V: 0, G: zeros, Delta: 0})
+		hits.N = zeros
+	}
+	if zeros := errDDSketch.zeros; zeros > 0 {
+		errors.Entries = append(errors.Entries, Entry{V: 0, G: zeros, Delta: 0})
+		errors.N = zeros
+	}
+	indexes := getIndexes(okDDSketch, errDDSketch)
+	for _, index := range indexes {
+		gErr := errDDSketch.count(index)
+		gHits := okDDSketch.count(index) + gErr
+		if gHits == 0 {
+			// gHits == 0 implies gErr == 0
+			continue
+		}
+		hits.N += gHits
+		v := okDDSketch.mapping.Value(index)
+		hits.Entries = append(hits.Entries, Entry{
+			V:     v,
+			G:     gHits,
+			Delta: int(2 * EPSILON * float64(hits.N-1)),
+		})
+		if gErr == 0 {
+			continue
+		}
+		errors.N += gErr
+		errors.Entries = append(errors.Entries, Entry{
+			V:     v,
+			G:     gErr,
+			Delta: int(2 * EPSILON * float64(errors.N-1)),
+		})
+	}
+	if hits.N > 0 {
+		hits.Entries[0].Delta = 0
+		hits.Entries[len(hits.Entries)-1].Delta = 0
+	}
+	if errors.N > 0 {
+		errors.Entries[0].Delta = 0
+		errors.Entries[len(errors.Entries)-1].Delta = 0
+	}
+	hits.compress()
```

Comment:
I think we can remove compression step. ddsketch already controls and reduce the size of the sketches.
Not compressing would a allow a conversion back to ddsketch without any loss. Allowing services to consider these as ddsketches or manually compress